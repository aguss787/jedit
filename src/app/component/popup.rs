use ratatui::layout::{Constraint, Flex, Layout, Rect};

pub fn popup_area(area: Rect, h: u16, w: u16) -> Rect {
    let vertical = Layout::vertical([Constraint::Length(h)]).flex(Flex::Center);
    let horizontal = Layout::horizontal([Constraint::Length(w)]).flex(Flex::Center);
    let [area] = vertical.areas(area);
    let [area] = horizontal.areas(area);
    area
}
