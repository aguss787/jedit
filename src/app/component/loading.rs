use std::time::Instant;

use super::popup::popup_area;
use ratatui::{
    layout::Rect,
    prelude::Buffer,
    text::Text,
    widgets::{Block, Clear, Padding, Widget},
};

pub struct Loading(Instant);

impl Default for Loading {
    fn default() -> Self {
        Self::new()
    }
}

impl Loading {
    pub fn new() -> Self {
        Loading(Instant::now())
    }

    fn loading_text(&self) -> Text {
        let elapsed = (self.0.elapsed().as_secs() % 4) as usize;
        Text::from(String::from_iter(
            "Loading".chars().chain(std::iter::repeat_n('.', elapsed)),
        ))
        .left_aligned()
    }
}

impl Widget for &Loading {
    fn render(self, area: Rect, buf: &mut Buffer)
    where
        Self: Sized,
    {
        let block = Block::bordered().padding(Padding::symmetric(1, 1));
        let area = popup_area(area, 5, 14);
        let inner_area = block.inner(area);

        Clear.render(inner_area, buf);

        block.render(area, buf);
        self.loading_text().render(inner_area, buf);
    }
}

#[cfg(test)]
mod test {
    use std::time::Duration;

    use insta::assert_snapshot;

    use crate::app::component::test_render::render_to_string;

    use super::*;

    #[test]
    fn render_test() {
        for i in 0..5 {
            let loading = Loading(Instant::now() - Duration::from_secs(i));
            assert_snapshot!(render_to_string(&loading));
        }
    }
}
