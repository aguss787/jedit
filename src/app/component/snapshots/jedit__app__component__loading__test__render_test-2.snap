---
source: src/app/component/loading.rs
expression: render_to_string(&loading)
---
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                 ┌────────────┐                                 "
"                                 │            │                                 "
"                                 │ Loading.   │                                 "
"                                 │            │                                 "
"                                 └────────────┘                                 "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
